Source/DafnyCore/AST/DafnyAst.cs

```diff
@@ -38,6 +40,36 @@ public abstract class INode {
     /// </summary>
     public abstract IEnumerable<INode> Children { get; }
 
+    public ISet<INode> Visit(Func<INode, bool> beforeChildren = null, Action<INode> afterChildren = null) {
+      beforeChildren ??= node => true;
+      afterChildren ??= node => { };
+
+      var visited = new HashSet<INode>();
+      var toVisit = new Stack<INode>();
+      toVisit.Push(this);
+      while (toVisit.Any()) {
+        var current = toVisit.Pop();
+        if (!visited.Add(current)) {
+          continue;
+        }
+
+        if (!beforeChildren(current)) {
+          continue;
+        }
+
+        foreach (var child in current.Children.Reverse()) {
```

Comment:
Most nodes have iterators (not lists) as children, so `Reverse` is probably not the most efficient.  Does the traversal order matter?  If so maybe you could use a dequeue instead of a stack with `Reverse`?

Reply:
Removed the Reverse for now. Actually seems this is only used in the follow-up PR, woops!